text/0082-dynamic-size-region.md

```diff
@@ -0,0 +1,129 @@
+we only discuss region size here.
+
+## Detailed design
+
+To reduce the count of regions, we can set the size of region to a larger value. More precisely, instead
+of 96MiB, we set the max region size to 10GiB (before compression). To get around the problem of
+concurrency, we introduce new concept bucket that is smaller ranges within a region. To make hotspot be
+recoginized and scheduled quickly, we choose the make the size of a region dynamic.
+
+![dynamic size](../media/dynamic-size.png)
+
+### Dynamic size
+
+The hotter a region is, the smaller its size becomes. To make it simplified, we choose 512MiB for hot regions
+and 10GiB for cold regions. So there are two types of split, hotspot split and general size split. Hotspot
+split is triggered by PD. General size split is triggered by TiKV itself.
+
+10GiB is just an example, it's allowed to change to a bigger or smaller value.
+
+### Bucket
+
+A region is split into several buckets logically. We will collect query stats by buckets and report the bucket
+to PD. For hotspot regions, buckets are split by scan. For cold regions, buckets are split by approximate
+range size. Every bucket should have the size about 128MiB. Their ranges and stats are all reported to PD. PD
+can detect hot buckets from all regions and decide whether need to split new hotspots and schedules them to
+balance.
+
+![dynamic size buckets](../media/dynamic-size-buckets.png)
+
+### Read and Write
+
+Since regions’ sizes become dynamic, region count in a single instance is not a problem any more.
+Hibernate regions can be deprecated, and lease renew by heartbeats can be enabled by default for
+all regions. This can bring significant improvement for read tail latency.
+
+After the region becomes larger, the chance to hit 1PC becomes higher. And the RPC count in a
+transaction can also be reduced. However, the hotspot chance is also increased. For writes, we rely
+on PD to split the hotspot in time and increase concurrency. In addition, we can also use unorder
+apply to make single region apply logs faster. Since unorder apply is a standalone feature, I’m not
+going into details here.
+
+For read hotspots, split should be triggered by PD, which can utilize the global statics from all
+regions and nodes. For normal read requests, TiKV will need to split its range into smaller buckets
+according to statics to increase concurrency. When TiDB wants to do a scan, it sends the RPC once,
+and TiKV will split the requests into smaller concurrent jobs.
+
+In the past design, follower read has also been discussed to offload works from leader. But TiDB
+can’t predict the progress of a follower, so latency can also become unpredictable. It’s more
+controllable to let PD split hotspots and schedule to get balance.
+
+Now that a region can be in 10GiB, a full scan can make the response exceed the limit of gRPC,
+which is 4GiB. So instead of unary, we need to use server streaming RPC to return the response as
```

Comment:
> we need to user server streaming RPC
Is it confirmed or just a candidate solution? An alternative is make clients send ranges instead of Region, maybe we can discuss about which is better.

Reply:
It's a candidate solution. I prefer client sends a unary request with the whole range. It's simple and efficient. Otherwise, TiDB may waste CPU at batch RPCs. A similar problem can be found here: https://github.com/tikv/client-go/issues/342.

Reply:
But streaming API is deprecated in TiDB

Reply:
It would be much easier to retry if TiDB split the request into buckets, each bucket can be sent concurrently.

If there is leader change, a single bucket request can be send to the new leader, the previous succeed bucket requests don't need to retry.

Reply:
If the request is split in TiKV side, snapshot is retrieved once, so there is no such case that retry some of the buckets due to leader change.

Reply:
If TiKV is slow, and TiDB encounter frequently time out, retry the whole region request would make the situation much worse.

Reply:
Then TiDB can change the behavior of retrying, it can know what ranges are missing easily.

Reply:
I left unary vs streaming as an unresolved question and will make decision by evaluation on maintainability and performance.